File: algoexpert/easy/sortedSquaredArray/python/sorted-squared-array.py
# solution1
# Time: O(nlogn)
# Space: O(n)
def sortedSquaredArray1(array):
    sortedSquares = [0 for _ in array]

    for idx in range(len(array)):
        value = array[idx]
        sortedSquares[idx] = value * value
    sortedSquares.sort()
    return sortedSquares


# solution2
# Time: O(n)
# Space: O(n)
def sortedSquaredArray2(array):
    sortedSquares = [0 for _ in array]
    smallerValueIdx = 0
    largerValueIdx = len(array) - 1

    for idx in reversed(range(len(array))):
        smallerValue = array[smallerValueIdx]
        largerValue = array[largerValueIdx]

        if abs(smallerValue) > abs(largerValue):
            sortedSquares[idx] = smallerValue * smallerValue
            smallerValueIdx += 1
        else:
            sortedSquares[idx] = largerValue * largerValue
            largerValueIdx -= 1

    return sortedSquares


result1 = sortedSquaredArray1([-2, -1])
result2 = sortedSquaredArray2([-2, -1])
print(result1, result2)

File: algoexpert/easy/sortedSquaredArray/typescript/sorted-squared-array.ts
// solution1
function sortedSquaredArray1(array: number[]): number[] {
  const sortedSquares = new Array<number>(array.length).fill(0);

  for (let idx = 0; idx < array.length; idx++) {
    const value = array[idx];
    sortedSquares[idx] = value * value;
  }
  sortedSquares.sort((a, b) => a - b);
  return sortedSquares;
}

// solution2
function sortedSquaredArray2(array: number[]): number[] {
  const sortedSquares = new Array<number>(array.length).fill(0);
  let smallerValueIdx = 0;
  let largerValueIdx = array.length - 1;

  for (let idx = array.length - 1; idx >= 0; idx--) {
    const smallerValue = array[smallerValueIdx];
    const largerValue = array[largerValueIdx];

    if (Math.abs(smallerValue) > Math.abs(largerValue)) {
      sortedSquares[idx] = smallerValue * smallerValue;
      smallerValueIdx++;
    } else {
      sortedSquares[idx] = largerValue * largerValue;
      largerValueIdx--;
    }
  }

  return sortedSquares;
}

const result1 = sortedSquaredArray1([1, 2, 4, 5, 7]);
const result2 = sortedSquaredArray2([9, 2, 4, 5, 7]);
console.table({ result1, result2 });
